controlplane/kubeadm/controllers/helpers.go

```diff
@@ -241,8 +243,35 @@ func (r *KubeadmControlPlaneReconciler) generateMachine(ctx context.Context, kcp
 		},
 	}
 
+	clusterConfig, err := json.Marshal(kcp.Spec.KubeadmConfigSpec.ClusterConfiguration)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal cluster configuration")
+	}
+	machine.SetAnnotations(map[string]string{controlplanev1.KubeadmClusterConfigurationAnnotation: string(clusterConfig)})
+
 	if err := r.Client.Create(ctx, machine); err != nil {
-		return errors.Wrap(err, "Failed to create machine")
+		return errors.Wrap(err, "failed to create machine")
 	}
 	return nil
 }
+
+// MachinesNeedingRollout return a list of machines that need to be upgraded.
+func (r *KubeadmControlPlaneReconciler) MachinesNeedingRollout(ctx context.Context, c *internal.ControlPlane) internal.FilterableMachineCollection {
+	now := metav1.Now()
+
+	// Ignore machines to be deleted.
+	machinesToUpgrade := c.Machines.Filter(machinefilters.Not(machinefilters.HasDeletionTimestamp))
+
+	// Return machines if their creation time are older than the KCP UpgradeAfter along with the machines that do not match KCP config.
+	if c.KCP.Spec.UpgradeAfter != nil && c.KCP.Spec.UpgradeAfter.Before(&now) {
+		return machinesToUpgrade.Filter(machinefilters.Or(
+			// Machines that are old.
+			machinefilters.OlderThan(c.KCP.Spec.UpgradeAfter),
+			//Machines that do not match with KCP config.
+			machinefilters.Not(machinefilters.MatchesKCPConfiguration(ctx, r.Client, *c.KCP, *c.Cluster)),
```

Comment:
AFAIK UpgradeAfter can be used to force-trigger a rolling upgrade, no matter of the configuration is changed or not. From the field description:
```
// UpgradeAfter is a field to indicate an upgrade should be performed
// after the specified time even if no changes have been made to the
// KubeadmControlPlane
```

Reply:
Right, these 2 filters are `Or`ed. I think it will trigger an upgrade even if there is no changes.

Reply:
wondering if we can avoid `if c.KCP.Spec.UpgradeAfter != nil && c.KCP.Spec.UpgradeAfter.Before(&now)`, but this is a nit

Reply:
If we remove `c.KCP.Spec.UpgradeAfter.Before(&now)`, all machines (existing or new created) will be rolled out each time we hit this method, if UpgradeAfter is sometime in the future.